haskell/test/SMT/Common/Gen.hs

```diff
@@ -0,0 +1,431 @@
+  addrFreezeHistoryList <- vectorOf poolSize genAddressFreezeHistory
+  let freezeHistory = zip addrs addrFreezeHistoryList
+  pure $ Map.fromList freezeHistory
+
+genQuorumThresholdAtCycle :: Natural -> GeneratorT QuorumThresholdAtCycle
+genQuorumThresholdAtCycle staked = do
+  quorumThreshold <- Gen.integral (Range.constant 1 3)
+  lvl <- get <&> gsLevel
+  pure $ QuorumThresholdAtCycle lvl quorumThreshold staked
+
+genModelState :: SmtOption -> GeneratorT (Address -> Address -> ModelState)
+genModelState SmtOption{..} = do
+
+  mkStore <- genStorage
+  config <- genConfig
+  let mkFs = \guardian gov -> FullStorage' (mkStore guardian gov) config
+
+  selfAddrPlaceholder <- genAddress
+  pure $ \guardian gov ->
+    let fs = mkFs guardian gov
+    in ModelState
+        { msFullStorage = fs
+        , msMutez = toMutez 0
+
+        , msLevel = fs & fsStorage & sStartLevel
+        , msChainId = dummyChainId
+        , msSelfAddress = selfAddrPlaceholder -- This will be replace when dao is originated
+        , msGovernanceTokenAddress = gov
+        , msContracts = mempty
+        , msProposalCheck = soProposalCheck
+        , msRejectedProposalSlashValue = soRejectedProposalSlashValue
+        , msDecisionLambda = soDecisionLambda
+        , msCustomEps = soCustomEps
+        }
+
+-------------------------------------------------------------
+-- Entrypoints
+-------------------------------------------------------------
+
+-- | Generate completely random calls
+genRandomCalls :: GeneratorT (ModelInputArg -> [ModelCall])
+genRandomCalls = do
+  addrs <- get <&> gsAddresses <<&>> fst
+  sender1 <- Gen.element addrs
+
+  -- gen entrypoints
+  calls <- genProposingProcess
+  updateDelegateCall <- genUpdateDelegate <&> mkSimpleContractCall sender1
+  transferOwnershipCall <- genTransferOwnership <&> mkSimpleContractCall sender1
+  acceptOwnershipCall <- genAcceptOwnership <&> mkSimpleContractCall sender1
+  mkTransferContractCall <- genTransferContractTokensCall sender1
+
+  pure $ \args@ModelInputArg{..} ->
+    [ updateDelegateCall
+    , transferOwnershipCall
+    , acceptOwnershipCall
+    , mkTransferContractCall miaGovAddr
+    ] <> (calls args)
+
+  where
+    genTransferContractTokensCall :: Address -> GeneratorT (Address -> ModelCall)
+    genTransferContractTokensCall sender1 = do
+      f <- genTransferContractTokens
+      pure $ \govAddr -> mkSimpleContractCall sender1 $ f govAddr
+
+
+mkSimpleContractCall :: Address -> Parameter -> ModelCall
+mkSimpleContractCall sender1 param = do
+  ModelCall
+    { mcAdvanceLevel = Nothing
+    , mcParameter = param
+    , mcSource = ModelSource
+        { msoSource = sender1
+        , msoSender = sender1
+        }
+    }
+
+genProposingProcess :: GeneratorT (ModelInputArg -> [ModelCall])
+genProposingProcess = do
+  userAddrs <- get <&> gsAddresses
+  sender1 <- fst <$> Gen.element userAddrs
+  delegate1 <- fst <$> Gen.element userAddrs
+  invalidFrom <- fst <$> Gen.element userAddrs
```

Comment:
Same problem here. The proposal generators simply draw one of these as random, e.g.:

```hs
-- From: SMT.BaseDAO.genPropose
from <- Gen.element [senderInput, invalidFrom, delegate1]
```

So this is effectively the same as simply drawing a single random address from the pool

Reply:
Nevermind, I see the flaw in my logic now. Below, `sender1` and `delegate` are setup in a different way, and `invalidFrom` is not setup at all. So yeah, this makes sense, my bad :sweat_smile:

Reply:
Yeah, they are setup in a different way.